prusti-viper/src/encoder/procedure_encoder.rs

```diff
@@ -5426,50 +5676,150 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
         &mut self,
         local: mir::Local,
         projection: &[mir::PlaceElem<'tcx>],
+        encode_kind: ArrayAccessKind,
     ) -> EncodingResult<(vir::Expr, Vec<vir::Stmt>, ty::Ty<'tcx>, Option<usize>)> {
         let (encoded_place, ty, variant_idx) = self.mir_encoder.encode_projection(local, projection)?;
-        let (encoded_expr, encoding_stmts) = self.postprocess_place_encoding(encoded_place)?;
+        let (encoded_expr, encoding_stmts) = self.postprocess_place_encoding(encoded_place, encode_kind)?;
 
         Ok((encoded_expr, encoding_stmts, ty, variant_idx))
     }
 
+    fn encode_array_lookup_shared(
+        &mut self,
+        base: PlaceEncoding<'tcx>,
+        index: vir::Expr,
+        array_ty: ty::Ty<'tcx>,
+    ) -> EncodingResult<(vir::Expr, Vec<vir::Stmt>)> {
+        let at = self.encoder.encode_array_types(array_ty)?;
+
+        let lookup_res: vir::Expr = self.cfg_method.add_fresh_local_var(at.elem_ty.clone()).into();
+        let val_field = self.encoder.encode_value_field(at.elem_ty_rs);
+        let lookup_res_val_field = lookup_res.clone().field(val_field);
+
+        let (encoded_base_expr, mut stmts) = self.postprocess_place_encoding(base, ArrayAccessKind::Shared)?;
+        stmts.extend(self.encode_havoc_and_allocation(&lookup_res));
+
+        let usize_ty = self.encoder.env().tcx().mk_ty(ty::TyKind::Uint(ty::UintTy::Usize));
+        let idx_val_int = index.field(self.encoder.encode_value_field(usize_ty));
+
+        let lookup_pure_call = at.encode_lookup_pure_call(
+            encoded_base_expr.clone(),
+            idx_val_int,
+        );
+
+        stmts.push(vir::Stmt::Assert(
+            vir::Expr::predicate_access_predicate(at.array_pred, encoded_base_expr, vir::PermAmount::Read),
+            vir::Position::default(),
+        ));
+
+        stmts.push(vir::Stmt::Inhale(vir!{ [ lookup_pure_call ] == [ lookup_res_val_field ] }));
+        Ok((lookup_res, stmts))
+    }
+
+    fn encode_array_lookup_mut(
+        &mut self,
+        base: PlaceEncoding<'tcx>,
+        index: vir::Expr,
+        array_ty: ty::Ty<'tcx>,
+        _loan: Option<Borrow>,
+        location: mir::Location,
+    ) -> EncodingResult<(vir::Expr, Vec<vir::Stmt>)> {
+        let at = self.encoder.encode_array_types(array_ty)?;
+
+        let res: vir::Expr = self.cfg_method.add_fresh_local_var(at.elem_ty.clone()).into();
+        let val_field = self.encoder.encode_value_field(at.elem_ty_rs);
+        let res_val_field = res.clone().field(val_field);
+
+        let (encoded_base_expr, mut stmts) = self.postprocess_place_encoding(base, ArrayAccessKind::Mutable(None, location))?;
+
+        let usize_ty = self.encoder.env().tcx().mk_ty(ty::TyKind::Uint(ty::UintTy::Usize));
+        let idx_val_int = index.field(self.encoder.encode_value_field(usize_ty));
```

Comment:
```suggestion
        let idx_val_int = vir::Expr::snap_app(index);
```

Reply:
i run into [this assertion](https://github.com/viperproject/prusti-dev/blob/master/prusti-viper/src/encoder/foldunfold/requirements.rs#L355) if i try that.. :-/

Reply:
Ok, then this either needs to be patched manually:
```suggestion
        let idx_val_int = self.encoder.patch_snapshots(vir::Expr::snap_app(index))?;
```
Or just left as is.

Reply:
hm, manually patching sounds fair enough, i'll go with that.